src/dft.cpp

```diff
@@ -204,10 +204,15 @@ dft_chunk *fields::add_dft(component c, const volume &where, const double *freq,
     double freq_max = 0;
     for (size_t i = 0; i < Nfreq; ++i)
       freq_max = std::max(freq_max, std::abs(freq[i]));
-    if ((freq_max > 0) && (src_freq_max > 0))
+    if ((freq_max > 0) && (src_freq_max > 0) && !has_nonlinearities(false))
```

Comment:
Calling `fields::has_nonlinearities(bool parallel)` with a `false` argument means that the `or_to_all` statement in `fields.cpp:670` (below) is never invoked. That would mean `fields::has_nonlinearities(bool parallel)` returns different values for different chunks. This would seem to be a bug.

Reply:
No, because of the subsequent `min_to_all` statement.

Reply:
If I replace `!has_nonlinearities(false)` with `!has_nonlinearities(true)`, `test_adjoint_solver` deadlocks when run with two chunks (as reported previously). This seems to be due to the `or_to_all` call which currently is never invoked. 

It's not obvious why `or_to_all` triggers a deadlock but `min_to_all` does not.

Reply:
The reason why it deadlocks is that `has_nonlinearities` here only gets executed if `(freq_max > 0) && (src_freq_max > 0)` is true, because `&&` is a [short-circuiting operator](https://stackoverflow.com/questions/45848858/what-is-short-circuit-evaluation-in-c).    

In the DFT adjoint simulation, where the adjoint sources get placed with `add_srcdata`, which is only called on chunks where the sources are present, some of the chunks are missing information about the sources and have `src_freq_max == 0`.   In consequence, those chunks are not calling `has_nonlinearities`, so it deadlocks if `parallel == true` and `or_to_all` is called.

In contrast, the `min_to_all` is always called on *every* chunk, because it is outside of these conditionals, so it does not deadlock.

Moreover, this means that the new `min_to_all` potentially fixes bugs even in linear simulations, since it was previously possible that some linear simulations over-estimated the `decimation_factor` on chunks for which no sources were present.

Reply:
I think it would have made the code more readable if, inside the block `if (decimation_factor == 0) { ... }` we had simply inserted:
```cpp
if has_nonlinearities()
   decimation_factor = 1
else {
  // set decimation_factor automatically...
}
```
The `bool` argument to `fields::has_nonlinearities()` is not necessary.

Reply:
The reason I added the `bool` argument is that we need to call `min_to_all` *anyway*, even for linear simulations, to make sure we don't over-estimate the `decimation_factor` in cases where a chunk is missing the source term present only in another chunk.   In that case, doing `or_to_all` is redundant, and is not cheap because it is a second MPI reduction.